prac_07/project_management.py

```diff
@@ -0,0 +1,233 @@
+    date_to_filtered_project = {}
+
+    print(f"----Showing projects from {filter_date_string}----")
+    filter_date = datetime.datetime.strptime(filter_date_string, '%d/%m/%Y').date()
+
+    for project in projects:
+        project_date = datetime.datetime.strptime(project.start_date, '%d/%m/%Y').date()
+
+        if project_date >= filter_date:
+            filtered_projects.append(project)
+            date_to_filtered_project[project_date] = project
+
+    for date, filtered_project in sorted(date_to_filtered_project.items()):
+        print(filtered_project)
+
+
+def get_valid_project_details():
+    """Get valid project details."""
+    print("Let's add a new project")
+    name = input("Name: ")
+    while name == "":
+        print("Project name can not be empty")
+        name = input("Name: ")
+
+    start_date = get_valid_date_string()
+
+    is_valid_priority = False
+    while not is_valid_priority:
+        try:
+            priority = int(input("Priority (1-10): "))
+            if priority > 10 or priority < 1:
+                print("1 (Highest priority) - 10 (Lowest priority")
+            else:
+                is_valid_priority = True
+        except ValueError:
+            print("Priority must be an integer (1-10)")
+
+    is_valid_cost_estimate = False
+    while not is_valid_cost_estimate:
+        try:
+            cost_estimate = float(input("Cots estimate: $"))
+            if cost_estimate < 0:
+                print("It must cost you something or must be free (0)")
+            else:
+                is_valid_cost_estimate = True
+        except ValueError:
+            print("Must be a number")
+
+    is_valid_completion_percentage = False
+    while not is_valid_completion_percentage:
+        try:
+            completion_percentage = float(input("Percent complete: "))
+            if completion_percentage < 0 or completion_percentage > 100:
+                print("Must be 0-100")
+            else:
+                is_valid_completion_percentage = True
+        except ValueError:
+            print("Percent must be a number")
+
+    return name, start_date, priority, cost_estimate, completion_percentage  # No problem with warning
+
+
+def update_project(projects):
+    """Update a Project completion percentage and/or priority."""
+    for i, project in enumerate(projects):
+        print(f"{i} {project}")
+    is_valid_project_choice = False
+    while not is_valid_project_choice:
+        try:
+            project_choice = int(input("Project choice: "))
+            chosen_project = projects[project_choice]
+
+            if project_choice < 0:
+                print("No such thing!")
+            else:
+                is_valid_project_choice = True
+        except IndexError:
+            print("That project not in the list yet!")
+        except ValueError:
+            print("Must be a number!")
+    print(chosen_project)  # No problem with warning
+
+    # Leave blank to retain existing values
+    try:
```

Comment:
Nice implementation of the error handling and optional inputs here. But should we reject the input if we entered random characters here?